### DISCUSSION TITLE
How to render text mode like in 0.74

### DISCUSSION THREAD

**Discussion by pglizniewicz** [dosbox-staging/dosbox-staging#4043] [Q&A] [answered]
What combination of options should I use to disable scalling / apect ratio correction? What I basically want is a 1:1 mapping of "game pixels" to display pixels.

**Comment by johnnovak** [comment]
You literally want 320x200 native pixels for a 320x200 game? We support no such thing.

Also, look up the help of the `aspect` settings.

**Reply by pglizniewicz** [reply]
What I really want is more like 720x400 in text mode in a window (1440x800 with 2x2 pixels will also work). I just want TUI programs without the CRT shaders and without odd looking characters due to scalling.

**Reply by interloper98** [reply]
Try:

```ini
[render]
glshader = sharp
integer_scaling = vertical
```

Drag to resize the window to your desired size, and then look at the console log to see the resulting resolution. For example:

```
DISPLAY: VGA 720x400 16-colour text mode 03h at 70.087 Hz throttled VFR,
 scaled to 1067x800 pixels with 1:1.35 (20:27) pixel aspect ratio
```

Then take that **_"scaled to [resolution]"_** and set that in your conf (for example):

```ini
[sdl]
windowresolution = 1067x800
```

Should get you most of the way, I think?

**Reply by johnnovak** [reply]
> What I really want is more like 720x400 in text mode in a window (1440x800 with 2x2 pixels will also work). I just want TUI programs without the CRT shaders and without odd looking characters due to scalling.

You could try `integer_scaling = vertical` as @interloper98 said, or you could set `aspect = off` (or `square-pixels`, it's the same thing).

Read the config descriptions, man, we wrote them for a reason 😉 

I'll do it for you now, once 😄 (with the CRT shader on, of course 😎 🤘🏻 )

<img width="1025" alt="image" src="https://github.com/user-attachments/assets/b4b8e97f-53d4-469a-921c-4291fe2563bd">

So, putting it together:

```ini
[render]
aspect = square-pixels
integer_scaling = vertical
```

With this, you'll always get exact integer multiple scaling factors, so 100% sharp pixels—as sharp as it gets 😆 Btw, with square pixels, you could also use `integer_scaling = horizontal`; it doesn't matter (look up its help text for details).

**Comment by pglizniewicz** [comment]
I wouldn't ask here before RTFM.

I've installed the old 0.74-3 for comparision. It renders 720x400 at 640x400 and that's what I'm looking for.

In dosbox-staging I've started with `glshader = sharp`. It didn't work as I've expected:

```
DISPLAY: VGA 720x400 16-colour text mode 03h at 70,087 Hz throttled VFR, scaled to 1067x800 pixels with 1:1,35 (20:27) pixel aspect ratio
```

I've added `aspect = off`. It made it worse:

```
DISPLAY: VGA 720x400 16-colour text mode 03h at 70,087 Hz throttled VFR, scaled to 1067x593 pixels with 1:1 (1:1) pixel aspect ratio
```

Adding `windowresolution = 640x400` didn't help either, I really don't understand how this parameter works, it never results in 640x400. Combining it with `aspect = off`, `integer_scalling = vertical` didn't help either. I really just wan't to make it look like in DOSBox 0.74.

**Reply by johnnovak** [reply, marked as answer]
> I wouldn't ask here before RTFM.

Well, I believe you, but you've missed something:

```
vga_8dot_font: Use 8-pixel-wide fonts on VGA adapters (disabled by default).
```

(Sorry, I just had to rub it in 😉)

> I really just wan't to make it look like in DOSBox 0.74.

`vga_8dot_font` will give you 640x480 (8 dot chars * 80 columns = 640) on VGA. That's effectively forcing EGA fonts on a VGA card, and original DOSBox never supported proper VGA fonts with the default config (probably they didn't care about it; "it's all about the games, man 😎" — and yeah, they're right). Maybe in `machine = vga_only` you got the proper 9px fonts, but I can't remember exactly...

If you want 8px fonts for the sake of familiarity, sure, go for it. Btw, some really early VGA cards also had 8px wide fonts, so it can be even somewhat authentic (but not with the S3 cards).

The S3 Trio64 we're emulating, though, has 9px-wide fonts, just like the vast majority of VGA cards. So we've improved things to emulate (S)VGA boards more authentically.

Read more about that in our release notes:
https://www.dosbox-staging.org/releases/release-notes/0.81.0/#accurate-vga-emulation

Your problem is that you want "sharp pixels", so that means no aspect ratio correction. But with no aspect correction, 9px wide fonts start looking a bit "fat", indeed. So, by forcing 8px wide fonts (EGA fonts) in conjunction with 1:1 PAR on a VGA card, you get closer to how things look on an actual CRT monitor...

Kinda... sorta... if you squint a bit 😉 

Anyway, here's your answer, and of course you might not care about authenticity at all, "just gimme how 0.74 looks like". So yeah, this is how to do that. Just realise you wanna emulate a shortcoming of 0.74 because of familiarity.

**Comment by pglizniewicz** [comment]
OK, I've done some testing, and I don't think I care enough about how text modes look like to fight with dosbox-staging, I will stick to defaults.

As for "authenticity", I become used to how comfortable modern terminals are, that personally, I would just render text modes like a modern terminal, with the ability to configure fonts and colors. But I guess this is out of scope for dosbox-staging.

**Reply by johnnovak** [reply]
> As for "authenticity", I become used to how comfortable modern terminals are, that personally, I would just render text modes like a modern terminal, with the ability to configure fonts and colors. But I guess this is out of scope for dosbox-staging.

Yes. Please familiarise yourself with our project goals (and non-goals):

https://www.dosbox-staging.org/about/#goals

TL;DR it's all about the games, then to some extent to emulate sitting in front of a real 80s/90s PC as closely as possible (as that's a powerful nostalgia trigger). Authenticity is very important for us to preserve the past.

If you want TTF fonts in text modes and other such modern features that did not exist on real MS-DOS, you can look into DOSBox X which has broader project goals and supports such enhancements.

Btw, you _can_ redefine the CLI colours via `cga_colors` iin Staging.
